src/OrchardCore/OrchardCore.Users.Core/Services/TwoFactorEmailTokenProvider.cs

```diff
@@ -0,0 +1,209 @@
+    /// </summary>
+    /// <param name="purpose">The purpose the token will be generated for.</param>
+    /// <param name="manager">The <see cref="UserManager{TUser}"/> that can be used to retrieve user properties.</param>
+    /// <param name="user">The user a token should be generated for.</param>
+    /// <returns>
+    /// The <see cref="Task"/> that represents the asynchronous operation, containing a constant modifier for the specified 
+    /// <paramref name="user"/> and <paramref name="purpose"/>.
+    /// </returns>
+    private static async Task<string> GetUserModifierAsync(string purpose, UserManager<IUser> manager, IUser user)
+    {
+        ArgumentNullException.ThrowIfNull(user);
+        var userId = await manager.GetUserIdAsync(user);
+
+        return $"Totp:{purpose}:{userId}";
+    }
+}
+
+/// <summary>
+/// The following code is influenced by <see href="https://github.com/dotnet/aspnetcore/blob/main/src/Identity/Extensions.Core/src/Rfc6238AuthenticationService.cs"/>
+/// </summary>
+internal sealed class Rfc6238AuthenticationService
+{
+    private static readonly UTF8Encoding _encoding = new(false, true);
+
+    private readonly TimeSpan _timeSpan;
+    private readonly TwoFactorEmailTokenLength _length;
+    private int? _modulo;
+
+    public Rfc6238AuthenticationService(TimeSpan timeSpan, TwoFactorEmailTokenLength length)
+    {
+        _timeSpan = timeSpan;
+        _length = length;
+    }
+
+    private int GetModuloValue()
+    {
+        // Number of 0's is length of the generated PIN.
+        _modulo ??= _length switch
+        {
+            TwoFactorEmailTokenLength.Two => 100,
+            TwoFactorEmailTokenLength.Three => 1000,
+            TwoFactorEmailTokenLength.Four => 10000,
+            TwoFactorEmailTokenLength.Five => 100000,
+            TwoFactorEmailTokenLength.Six => 1000000,
+            TwoFactorEmailTokenLength.Seven => 10000000,
+            TwoFactorEmailTokenLength.Eight or TwoFactorEmailTokenLength.Default => 100000000,
+            _ => throw new NotSupportedException("Unsupported token length.")
+        };
+
+        return _modulo.Value;
+    }
+
+    internal int ComputeTOTP(byte[] key, ulong timestepNumber, byte[] modifierBytes)
+    {
+        // See https://tools.ietf.org/html/rfc4226
+        // We can add an optional modifier.
+        Span<byte> timestepAsBytes = stackalloc byte[sizeof(long)];
+        var res = BitConverter.TryWriteBytes(timestepAsBytes, IPAddress.HostToNetworkOrder((long)timestepNumber));
+        Debug.Assert(res);
+
+        var modifierCombinedBytes = timestepAsBytes;
+        if (modifierBytes is not null)
+        {
+            modifierCombinedBytes = ApplyModifier(timestepAsBytes, modifierBytes);
+        }
+
+        Span<byte> hash = stackalloc byte[HMACSHA1.HashSizeInBytes];
+        res = HMACSHA1.TryHashData(key, modifierCombinedBytes, hash, out var written);
+        Debug.Assert(res);
```

Comment:
Why the assertion is needed?

Reply:
This code is mainly influenced by the references class. So it's just copied for the most part

Reply:
I thought that :)